### PR TITLE
feat: Persistent directories for docker

### PR DESCRIPTION
**Pull request by gbouv**
## Description:
Adds the docker backend to support persistent directories for user services.

## Is this change user facing?
NO

### CODE REVIEW

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/files_artifacts_expansion.go [hunk 1]
```diff
@@ -302,23 +302,25 @@ func createFilesArtifactsExpansionVolumes(
...
-					)
-					logrus.Warnf("You'll need to clean up volume '%v' manually!", volumeNameStr)
-				}
-			}
-		}()
-
 		result[mountpointExpanderWants] = volumeNameStr
 	}
+
+	defer func() {
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/files_artifacts_expansion.go:308]: nice!

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/persistent_directories.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+			)
+		}
+		result[volumeName] = serviceDirpath
+	}
+
+	defer func() {
+		if !shouldDeleteVolumes {
+			return
+		}
+		for _, volumeNameStr := range result {
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/persistent_directories.go]: `result` could contain previous existing volumes right? Is it ok to remove those also?

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go [hunk 3]
```diff
@@ -481,15 +482,51 @@ func createStartServiceOperation(
...
 							logrus.Errorf("Starting the service failed so we tried to delete files artifact expansion volume '%v' that we created, but doing so threw an error:\n%v", volumeName, err)
 							logrus.Errorf("You'll need to delete volume '%v' manually!", volumeName)
 						}
 					}
 				}
 			}()
-			volumeMounts = candidateVolumeMounts
+
+			for dirpath, volumeName := range candidateVolumeMounts {
+				if _, found := volumeMounts[dirpath]; found {
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/start_user_services.go:495]: nice!

File: container-engine-lib/lib/backend_impls/docker/object_attributes_provider/enclave_object_attributes_provider.go [hunk 4]
```diff
@@ -321,6 +331,56 @@ func (provider *dockerEnclaveObjectAttributesProviderImpl) ForSingleFilesArtifac
...
+		return nil, stacktrace.Propagate(err, "An error occurred getting labels for files artifact expansion volume with UUID '%v'", guidStr)
+	}
+
+	serviceUuidLabelValue, err := docker_label_value.CreateNewDockerLabelValue(serviceUuidStr)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "An error occurred creating a Docker label value from service GUID string '%v'", serviceUuidStr)
+	}
+	labels[label_key_consts.UserServiceGUIDDockerLabelKey] = serviceUuidLabelValue
+	labels[label_key_consts.VolumeTypeDockerLabelKey] = label_value_consts.PersistentDirectoryVolumeTypeDockerLabelValue
+	// TODO Create a KurtosisResourceDockerLabelKey object, like Kubernetes, and apply the "user-service" label here?
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_impls/docker/object_attributes_provider/enclave_object_attributes_provider.go:373]: is this TODO for later? or is this something missing on this PR?

Comment (leoporoli) [reply container-engine-lib/lib/backend_impls/docker/object_attributes_provider/enclave_object_attributes_provider.go:373]: I can image that it's for later but I'll ask just to double check

File: container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package service_directory
+
+type DirectoryPersistentKey string
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go:3]: nice!

**Review by leoporoli — APPROVED** [review]

File: container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/persistent_directories.go (hunk 2 above)
Comment (gbouv) [reply container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/user_services_functions/persistent_directories.go]: Yes you're right, I guess we should probable leave the already existing volume untouched in this case. I'll change that. Good catch!

File: container-engine-lib/lib/backend_impls/docker/object_attributes_provider/enclave_object_attributes_provider.go (hunk 4 above)
Comment (gbouv) [reply container-engine-lib/lib/backend_impls/docker/object_attributes_provider/enclave_object_attributes_provider.go:373]: I have to admit this is from a copy paste from `ForSingleFilesArtifactExpansionVolume` and I am not sure we want to do that. I'll leave it around for now. I think until we should not add this label unless we have a concrete use case for it. If you can think of one, I'll add :)

**Pull request merged**